components/brave_rewards/core/attestation/attestation.h

```diff
@@ -32,7 +33,7 @@ class Attestation {
                        ConfirmCallback callback) = 0;
 
  protected:
-  LedgerImpl* ledger_;  // NOT OWNED
+  const raw_ref<LedgerImpl> ledger_;
```

Comment:
I'm wondering if we actually want these `raw_ref<>`s to be `const` member variables. In general, that's an anti-pattern, as the type cannot be moved from, it implicitly deletes the move/copy-assignment operator, etc. I assume the intent here would be to prevent `ledger_` from being rebound? If so, then I think we'd be better off enforcing this via `const` member functions.

Reply:
I understand your concern, and I had also taken that into consideration myself. So let me clarify why adding these constant data members is the advisable approach.

What I'm doing here is in line with upstream practice. `raw_ref` will cause a crash if used after moved from. A type that holds a raw ref will require a custom move constructor anyways, if we want to support move operations in these types.

```c++
// # Use after move
//
// The raw_ref type will abort if used after being moved.
//
// # Constness
//
// Use a `const raw_ref<T>` when the smart pointer should not be able to rebind
// to a new reference. Use a `const raw_ref<const T>` do the same for a const
// reference, which is like `const T&`.
```
https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:base/allocator/partition_allocator/pointers/raw_ref.h

I particularly prefer types to implement move constructors, but for these types in question, I would rather not just do things for the sake of doing them, as in practice providing a move constructor is not an imperative. Furthermore, I would prefer to keep our code in line with what is practice upstream.

Reply:
I meant:
```
struct Struct {
  const int member = 42;
};

Struct s1;
s1 = Struct{};  // doesn't work
Struct s2;
s1 = s2;  // doesn't work
```

Reply:
I understand that. I'm saying that to have these types being non-copyable/non-moveable is actually not bad. Like I said, I'm in favour of having move constructors, but that would imply something else about these types that, as it seems to me, it is something we don't need.

To summarise: We do want to delete the move constructor, because a raw_ref cannot be accessed anymore once the object is moved from. Should a type support move, I'm inclined to review that, but I don't think we should worry about having copyability/move-ability for types we don't copy nor move.

Reply:
Interesting you brought that to my attention. I think you are right in part.

```c++
  // Disallow holding references to temporaries.
  raw_ref(const T&& p) = delete;
  raw_ref& operator=(const T&& p) = delete;
```

So on your example above
```
Struct s4(std::move(s1)); // defaults back to the copy constructor
```

So `Struct` has no move constructor, as it gets default deleted because `raw_ptr`'s one is deleted.

Reply:
Ah, you're right again. Thanks mate for the review, and your time looking into this.